src/tactic/fin_cases.lean

```diff
@@ -3,38 +3,75 @@ Copyright (c) 2018 Scott Morrison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Author: Scott Morrison
 
-Case bashing on `fin n`, for explicit numerals `n`.
+Case bashing:
+* on `x ∈ A`, for `A : finset α` or `A : list α`, or
+* on `x : A`, with `[fintype A]`.
 -/
-import data.fin
+import data.fintype
+import tactic.norm_num
 
 namespace tactic
 open lean.parser
 open interactive interactive.types expr
+open conv.interactive
 
-meta def fin_cases_at (e : expr) : tactic unit :=
-do `(fin %%n) ← infer_type e,
-   n ← eval_expr ℕ n,
-   [(_, [val, bd], _)] ← cases_core e [`val, `bd],
-   -- We now call `cases val` n times, rotating the generated goals out of the way.
-   iterate_at_most n (do val ← get_local `val, cases val, rotate 1),
-   -- We've got an absurd hypothesis `bd`, but it is messy: it looks like
-   -- `nat.succ (... (nat.succ val)) < n`
-   -- So we rewrite it as `bd : val + 1 + ... + 1 < n`, and use `dec_trivial` to kill it.
-   ss ← mk_const `nat.succ_eq_add_one,
-   bd ← get_local `bd,
-   (list.range n).mfoldl (λ bd _, do rewrite_hyp ss bd) bd,
-   to_expr ``(absurd %%bd dec_trivial) >>= exact,
-   -- We put the goals back in order, and clear the `bd` hypotheses.
-   iterate_exactly n (do rotate_right 1,
-                         try `[rw [fin.mk_val]],
-                         try (get_local `bd >>= clear))
+/-- Checks that the expression looks like `x ∈ A` for `A : finset α` or `A : list α`,
+    and returns the type α. -/
+meta def guard_mem_fin (e : expr) : tactic expr :=
+do t ← infer_type e,
```

Comment:
It would be nice if this handled `multiset` as well. Does it make sense to try to unify the type of `e` with `list`, `finset`, etc. (with a strict reducibility setting, perhaps) instead of looking for a syntactic match? I can't think of an obvious situation where this would be necessary, but I can imagine it could show up.

Reply:
I've added `multiset` as another case, for now. I'm not immediately sure how to do the test (and return in the inner type) via unification. Any hints?

Reply:
I'd create a sort metavariable `a`. Then try ``to_expr `(list %%a) >>= unify t`` and the other variants. If one succeeds, instantiate `a` and it will be your inner type. The syntax will be slightly different if you give `unify` a reducibility parameter.

Reply:
Grah... I've attempted this, and it seems to work ... but somehow later goals end up in the wrong order and break everything. I'm confused about what's going on!

I've left in, commented out, the block of code that uses unification here instead of syntactic matching. If someone wants to decipher what's wrong, great. Otherwise, I'm not convinced that this generalisation is worth more debugging, and would prefer to just remove the commented out stuff and merge as is.

Reply:
I can try to debug this on Monday. Using the optional arguments to `to_expr`, e.g. `subgoals := ff`, might make a difference.

Reply:
`subgoals := ff` did the trick, thank you! I've switched to testing that the type of `A` in `x \mem A` is a `finset/list/multiset` using unification rather than syntactic matching.